### PR TITLE
impl From<u32> for instructions enums

### PR DESCRIPTION
**Pull request by losynix**
Addresses issue https://github.com/capstone-rust/capstone-rs/issues/56 with your [solution 4](https://github.com/capstone-rust/capstone-rs/issues/56#issuecomment-455254742): Add a function to X86Insn (and other archs) to create an X86Insn from an InsnId.

Now mothran's [example](https://github.com/capstone-rust/capstone-rs/issues/56#issue-398715105) can be written like this:

```rust
fn is_valid_ret(insn: &Insn) -> bool {
    match X86Insn::from_insn_id(insn.id().0) {
        X86Insn::X86_INS_RET => true,
        _ => false
    }
}
```

Parsing and appending code to the generated bindings is a bit dirty, maybe Rust will have a good and safe way to cast integers to enums one day.

### CODE REVIEW

File: capstone-sys/build.rs [hunk 1]
```diff
@@ -148,6 +152,62 @@ fn env_var(var: &str) -> String {
...
+    for cap in re.captures_iter(&bindings) {
+        impl_arch_enum.push_str(
+            &format!("impl {}_insn {{\n
+            pub fn from_insn_id(id: u32) -> {}_insn {{\n
+            match id {{\n",
+            &cap["arch"], &cap["arch"]));
+
+        // find instructions and their id
+        let re2 = Regex::new(
+            &format!("{}_INS_(?P<ins>[A-Z]+) = (?P<id>\\d+)",
```
Comment (tmfink) [inline capstone-sys/build.rs]: You need to account for underscores, numbers, and possible after `_INS_`.
We want to have this be future-proof.

For example, you are missing `MIPS_INS_ABSQ_S`, `MIPS_INS_ADDIUR1SP`.

Reference: https://doc.rust-lang.org/stable/reference/identifiers.html

File: capstone-sys/build.rs [hunk 2]
```diff
@@ -148,6 +152,62 @@ fn env_var(var: &str) -> String {
...
+
+    for cap in re.captures_iter(&bindings) {
+        impl_arch_enum.push_str(
+            &format!("impl {}_insn {{\n
+            pub fn from_insn_id(id: u32) -> {}_insn {{\n
+            match id {{\n",
+            &cap["arch"], &cap["arch"]));
+
+        // find instructions and their id
+        let re2 = Regex::new(
```
Comment (tmfink) [inline capstone-sys/build.rs]: Use descriptive names, not `re2`.

**Comment by tmfink** [comment]
Thanks for the PR! I have avoided doing this kind of thing, but maybe is what we want.
I have made some review comments.

**Comment by losynix** [comment]
Thank you, I updated the code following your comments except for the `TryFrom` trait implementation which requires returning a `Result` instead of a `<arch>_INS_INVALID` variant, not sure what would be more comfortable to use ?

**Comment by tmfink** [comment]
> Thank you, I updated the code following your comments except for the `TryFrom` trait implementation which requires returning a `Result` (thus an `Error` instead of a `<arch>_INS_INVALID` variant), not sure what would be more comfortable to use ?

You can just use the `From` trait to convert directly, including the `<arch>_INS_INVALID`.

In the future, I think it will make sense to define a separate enum for each arch instructions without `<arch>_insns_invalid`. When conversions fail, they would be able to return `Option<ArchInsn>`.

**Comment by losynix** [comment]
> You can just use the `From` trait to convert directly, including the `<arch>_INS_INVALID`.

Sure, done.

> In the future, I think it will make sense to define a separate enum for each arch instructions without `<arch>_insns_invalid`. When conversions fail, they would be able to return `Option<ArchInsn>`.

It would make sense indeed.

**Comment by losynix** [comment]
everything good for you ?

File: capstone-sys/build.rs [hunk 3]
```diff
@@ -40,6 +40,15 @@ extern crate bindgen;
 
 extern crate cc;
 
+#[cfg(feature = "use_bindgen")]
+extern crate regex;
```
Comment (tmfink) [inline capstone-sys/build.rs]: I should upgrade capstone-sys to the 2018 edition so that you don't need to do this

**Comment by tmfink** [comment]
Thanks for the PR! Sorry for the wait.

**Comment by losynix** [comment]
no problem, thank you !

**Pull request merged**